### PR TITLE
Revive macOS CI tests

### PR DESCRIPTION
**Pull request by nilsvu**
## Proposed changes

Revive CI tests on macOS. Run on PRs, but don't require the tests to pass. The tests are useful for catching missing link dependencies, since macOS / AppleClang is stricter about those than Linux, and it's annoying when they break macOS builds on develop.

- [x] Shared libs on macOS work since #2680
- [x] Installation of dependencies on macOS with Spack is easier since #3712
- [x] Depends on #3773 

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### CODE REVIEW

File: docs/Installation/Installation.md [hunk 1]
```diff
@@ -19,7 +19,7 @@ installation_on_clusters "Installation on clusters" page.
 * [Git](https://git-scm.com/)
 * BLAS (e.g. [OpenBLAS](http://www.openblas.net))
 * [Blaze](https://bitbucket.org/blaze-lib/blaze/overview) v3.5
-* [Boost](http://www.boost.org/) 1.60.0 or later
+* [Boost](http://www.boost.org/) between 1.63.0 and 1.72.0
```
Comment (wthrowe) [inline docs/Installation/Installation.md]: Please don't make unrelated changes in your PRs.  This is only correct for OSX (or generally systems with case-insensitive file systems).  Higher versions work fine in other cases.

**Review by wthrowe — CHANGES_REQUESTED** [review]

File: docs/Installation/Installation.md (hunk 1 above)
Comment (nilsvu) [reply docs/Installation/Installation.md]: Sure, I dropped this commit for now and will move it to a separate PR.

File: .github/workflows/MacOs.yaml [hunk 2]
```diff
@@ -110,17 +109,19 @@ jobs:
...
         run: |
           cd $HOME/dependencies
-          wget http://charm.cs.illinois.edu/distrib/charm-${CHARM_VERSION}.tar.gz
-          tar -xzf charm-${CHARM_VERSION}.tar.gz
-          cd ./charm-v${CHARM_VERSION}
-          ./build charm++ multicore-darwin-x86_64 -j4 -O0 $SPECTRE_MIN_MACOS
-          rm ../charm-${CHARM_VERSION}.tar.gz
+          git clone https://github.com/UIUC-PPL/charm && cd charm
+          # This is the commit that fixed #2397
+          git checkout e7e9d4b5b27608b89ab09f021b6b5194267f74fe
```
Comment (nilsdeppe) [inline .github/workflows/MacOs.yaml]: I'm confused, why does this install an old version of Boost to avoid the issue, but then check out a random hash to avoid the boost issue... I feel like we should only check out the exact Charm++ version we support

Comment (nilsvu) [reply .github/workflows/MacOs.yaml]: #2397 has nothing to do with boost. Some `.tbd` files appeared for some reason on GitHub's latest macOS VM and Charm couldn't handle those files, so our CI build broke. This commit https://github.com/UIUC-PPL/charm/commit/e7e9d4b5b27608b89ab09f021b6b5194267f74fe is the one that fixes the issue. It will be in charm v6.11, but they'll probably not have another bugfix release before that. So I figured it's better to checkout the particular commit than to checkout whatever the latest `master` is when the CI build runs. It's the closest thing to a bugfix release we can do I think.

**Review by nilsdeppe — DISMISSED** [review]

**Comment by kidder** [comment]
marking as in progress until MacOS build passes

**Comment by nilsvu** [comment]
Yep, I'll wait with this until some of the recent build system PRs are in that make it easier to compile on macOS.

**Comment by nilsvu** [comment]
This is ready for review now. The macOS job takes ~1h using caches from `develop`, see https://github.com/nilsvu/spectre/runs/4996097358?check_suite_focus=true.

**Review by nilsdeppe — DISMISSED** [review]

**Comment by nilsvu** [comment]
@wthrowe or @kidder could you take a look at this one?

File: .github/workflows/Tests.yaml [hunk 3]
```diff
@@ -602,6 +602,158 @@ ${{ env.CACHE_KEY_SUFFIX }}"
...
+            -D STUB_EXECUTABLE_OBJECT_FILES=ON \
+            -D STUB_LIBRARY_OBJECT_FILES=ON \
+            -D USE_PCH=ON \
+            -D USE_CCACHE=ON \
+            -D SPECTRE_TEST_TIMEOUT_FACTOR=5 \
+            $GITHUB_WORKSPACE
+      - name: Build tests
+        working-directory: build
+        run: |
+          make -j4 RunTests
```
Comment (kidder) [inline .github/workflows/Tests.yaml]: I believe that only two are available so `-j2`

**Review by kidder — CHANGES_REQUESTED** [review]

**Comment by kidder** [comment]
@wthrowe you have a old requested change, so please look at this

File: cmake/AddInputFileTests.cmake [hunk 4]
```diff
@@ -5,6 +5,15 @@ spectre_define_test_timeout_factor_option(INPUT_FILE "input file")
 
 find_package(Python REQUIRED)
 
+# Environment variables for test
+set(_TEST_ENV_VARS "")
```
Comment (wthrowe) [inline cmake/AddInputFileTests.cmake:9]: What's the leading underscore about?

Comment (nilsvu) [reply cmake/AddInputFileTests.cmake:9]: This is how it's named in `SpectreSetupPythonPackage.cmake` 🤷‍♂️ Some people prefix local variables with leading underscores in CMake.

File: .github/workflows/Tests.yaml (hunk 3 above)
Comment (nilsvu) [reply .github/workflows/Tests.yaml]: squashed this in and rebased

**Review by wthrowe — APPROVED** [review]

**Review by kidder — APPROVED** [review]
